src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,368 @@
+            s = e
+
+    def datavector(self):
+        splits = [self.D_prime[:, self.feature_indices_map[attr]] for attr in self.domain.attrs]
+
+        letters = "bcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        assert(
+            len(letters) >= len(self.domain)
+        ), 'High dim domain'
+
+        input_dims = ','.join(f'a{letters[i]}' for i in range(len(splits)))
+        output_dims = ''.join(f'{letters[i]}' for i in range(len(splits)))
+        einsum_str = f'{input_dims}->a{output_dims}'
+
+        joint = jnp.einsum(einsum_str, *splits).mean(axis=0)
+        return np.array(joint).flatten()
+
+    def project(self, cl: tuple):
+        cols = list(itertools.chain.from_iterable(self.feature_indices_map[attr] for attr in cl))
+        D_proj = self.D_prime[:, cols]
+        domain_proj = self.domain.project(cl)
+        return ProjectableData(D_proj, domain_proj)
+    
+    def synthetic_data(self, num_samples, seed=0):
+        n, _ = self.D_prime.shape
+        key = jax.random.PRNGKey(seed)
+        key, sk = jax.random.split(key)
+        sampled_indices = jax.random.choice(key, n, shape=(num_samples,), replace=True)
+        sampled_data = self.D_prime[sampled_indices]
+
+        result = []
+        for attr in self.domain.attrs:
+            p = sampled_data[:, self.feature_indices_map[attr]]
+            key, sk = jax.random.split(key)
+            ids = jax.random.categorical(sk, jnp.log(jnp.clip(p, a_min=1e-30, a_max=1.0)), axis=-1)
+            result.append(ids)
+
+        sampled_df = pd.DataFrame(np.stack([np.array(r) for r in result], axis=1), columns = list(self.domain.attrs))
+        sampled_data = Dataset(sampled_df, self.domain)
+        return sampled_data
+
+
+def relaxed_projection_estimation(
+    domain: Domain,
+    measurements: List[tuple],
+    iters: int = 1000,
+    lr: float = 0.01,
+    D_start: jax.Array = None,
+    seed: int = 0,
+    D_size: int = 1000,
+) -> ProjectableData:
+    """
+    API function for relaxed projection mechanism, return a 
+    ProjectableData class, which can further be used to conduct 
+    projection and synthesize data.
+
+    Args:
+        domain: a Domain class.
+        measurements: a list of tuples, each of which should be in 
+                      form ``(clique, measurement of clique)``.
+        iters: number of iterations in training.
+        lr: learning rate in training.
+        D_start: initial data on which we start the optimization. If 
+                 None, the optimization will start from a randomly 
+                 initialized data.
+        seed: random seed.
+        D_size: number of records in optimizable data.
+    """
+    key = jax.random.PRNGKey(seed)
+    if D_start is None:
+        D_start = _initialize_synthetic_dataset(key, num_generated_points=D_size, data_dimension=np.sum(domain.shape))
+
+    stat_dim = _obtain_dim(measurements = measurements)
+    statistics = [MarginalStatistics(domain, dim) for dim in stat_dim]
```

Comment:
I think the MarginalStatistics class should not be necessary if you have a MarginalLossFn.  The pattern could be:

marginals = { cl : D.project(cl) for cl in cliques }
loss = loss_fn(marginals)

And this whole computation should be differentiable by jax